File: 647.java
class Solution {
    public int countSubstrings(String S) {
        int length = S.length(), result = 0;
        for (int i = 0; i < length; i++) {
            int left = i - 1, right = i;
            while (right < length - 1 && S.charAt(right) == S.charAt(right + 1))
                right++;
            result += (right - left) * (right - left + 1) / 2;
            i = right++;
            while (left >= 0 && right < length && S.charAt(right++) == S.charAt(left--))
                result++;
        }
        return result;
    }
}

File: Python/647.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        def palindrome(left, right):
            count = 0
            while 0 <= left and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1
                count += 1

            return count

        count = 0
        for i in range(len(s)):
            odd = palindrome(i - 1, i + 1)
            even = palindrome(i, i + 1)
            count += odd + even + 1

        return count
